Documentation/Communicational Patterns/2. Tooling/2.5. Environment.md

```diff
@@ -0,0 +1,20 @@
+![document type: vision](https://themomax.github.io/resources/markdown-labels/document_type_vision.svg)
+
+# Environment
+
+The `@Environment` property wrapper provides different functionality depending on the `wrappedValue`. For more detail on how the `wrappedValue` is provided to the environment and what values are provided by the Apodini framework refer to [PropertyWrapper -> RequestAndEnvironment](../../PropertyWrapper/RequestAndEnvironment.md).
+
+## ObservableObject
+
+If the `wrappedValue` conforms to `ObservableObject`, the `@Environment` behaves exactly the same as [`@ObservedObject`](./2.4.%20ObservedObject.md).
+
+In any other case `@Environment` has no influence on the `Handler`'s lifetime or control flow.
+
+## Connection
+
+Apodini provides a `Connection` through the environment. This object can be accessed using `@Environment(\.connection)`. The `Connection` provides useful protocol-agnostic information about the current state of the communication. The `Connection` could provide a `start: Date` so the service could e.g. close connections with an timeout error.
+
+The most important property of `Connection` is `state: ConnectionState`. This state can be `.open` or `.end`, where `.end` signalizes that the protocol expects the connection to be closed now. The `Handler` will be destructed after the next evaluation that results in a `.send(Request)`. Usage of this special environment object may cause exporters to utilize a client-side stream for implementation of the respective endpoint.
```

Comment:
Can `.end` also be used to close the connection to the client from the server or does it only signal when the client closes its connection? Something like `.end(Response)` where this would send the last response from the server and also closes the connection to the client. I think a similar behavior can be useful for server side stream.

Reply:
No, `.end` cannot be used to close the connection. I think that this is a relatively rare requirement. My idea was that usually a client would unsubscribe from a subscription gracefully, or would just close their connection to the server. Both events would be handled at an exporter-level. If a developer really wants to close a stream from the server, they have to cancel all `@Published`s the `Handler` is subscribed to.

Reply:
I agree with @awoc that a `.end(Response)` would be reasonable approach to indicate that a connection is closed from the web service and have suggested a similar a idea above that would also fit into the idea that @awoc provided.

Reply:
The initial counter-argument I saw with this approach was that you cannot have a silent endpoint, that never sends a service-message but only consumes requests (e.g. supported by WebSocket). Thus I created two new cases: .end and .final(Response).